### PR TITLE
Agfs fixed fee checkboxes

### PR DESCRIPTION
**Pull request by jsugarman**
#### What
AGFS fixed fee UI improvements

#### Ticket

[board ticket description](link)

#### Why
Follow new designs

#### How
Instatiate eligible fixed fees in controller and present each one under its checkbox fee type

--------

#### TODO (wip)

 - [x] persistence problems
    - [x] destroy when nil value
    - [x] destroy when not applicable to case type?
 - [x] fee calc rewiring
 - [x] error message rewiring
 - [x] change case type journey

### CODE REVIEW

File: app/presenters/error_presenter.rb [hunk 1]
```diff
@@ -31,23 +31,29 @@ def size
...
-        short_message = generate_standard_short_message(fieldname, error)
-        api_message = generate_standard_api_message(fieldname, error)
-      end
+      messages = populate_messages(emt, error, fieldname)
+      next if @error_details[fieldname] && @error_details[fieldname][0].long_message.eql?(messages[1])
       @error_details[fieldname] = ErrorDetail.new(fieldname,
-                                                  long_message,
-                                                  short_message,
-                                                  api_message,
+                                                  messages[1],
```
Comment (jsugarman) [inline app/presenters/error_presenter.rb]: could use a struct of messages for clarity

```
Message = Struct.new(:long, :short, :api)
```
and use:
```
message = messages(emt, error, fieldname)

def messages
  message = Message.new
  if ...
    message.long = emt.long_message
     etc...
 
    return message
```

and use
```
 ErrorDetail.new(fieldname, message.long, etc...)
```

File: features/page_objects/sections/fixed_fee_section.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+  section :adjourned_appeals_committals_and_breaches, FeeSection, "#adjourned-appeals-committals-and-breaches > .fixed-fee-group"
+  section :appeals_to_the_crown_court_against_sentence, FeeSection, "#appeals-to-the-crown-court-against-sentence > .fixed-fee-group"
+  section :appeals_to_the_crown_court_against_conviction, FeeSection, "#appeals-to-the-crown-court-against-conviction > .fixed-fee-group"
+  section :number_of_cases_uplift, FixedFeeCaseNumbersSection, "#number-of-cases-uplift > .fixed-fee-group"
+  section :number_of_defendants_uplift, FeeSection, "#number-of-defendants-uplift > .fixed-fee-group"
+  section :standard_appearance_fee, FeeSection, "#standard-appearance-fee > .fixed-fee-group"
+  section :elected_case_not_proceeded, FeeSection, "#elected-case-not-proceeded > .fixed-fee-group"
+
+ #  sections :group, '.fixed-fee-group' do
```
Comment (jsugarman) [inline features/page_objects/sections/fixed_fee_section.rb]: remove

File: features/page_objects/claim_form_page.rb [hunk 3]
```diff
@@ -56,13 +59,14 @@ class ClaimFormPage < SitePrism::Page
   element :continue_button, 'div.button-holder > input.button.left'
 
   section :basic_fees, BasicFeeSection, "div#basic-fees"
+  section :fixed_fees, FixedFeeSection, "div#fixed-fees"
+
+  # TODO :remove
```
Comment (jsugarman) [inline features/page_objects/claim_form_page.rb]: remove

File: features/step_definitions/advocate_claim_steps.rb [hunk 4]
```diff
@@ -119,11 +126,24 @@
...
 
+Then(/^I add a '(.*?)' fixed fee with case numbers$/) do |name|
+  @claim_form_page.fixed_fees.check(name)
+  wait_for_ajax
+  @claim_form_page.fixed_fees.fee_block_for(name).case_numbers.set "T20170001"
+  @claim_form_page.fixed_fees.set_quantity(name)
+  wait_for_ajax
+end
+
 When(/^I set the last fixed fee value to '(.*?)'$/) do |value|
```
Comment (jsugarman) [inline features/step_definitions/advocate_claim_steps.rb]: possibly no longer required?

File: app/controllers/external_users/advocates/claims_controller.rb [hunk 5]
```diff
@@ -18,7 +18,8 @@ def build_nested_resources
   end
 
   def build_fixed_fees
-    existing_fixed_fee_fee_type_ids = @claim.fixed_fees.map(&:fee_type_id)
+    # NOTE: the `reload` is !important. prevents duplicate fees being created during error flows
+    existing_fixed_fee_fee_type_ids = @claim.fixed_fees.reload.map(&:fee_type_id)
```
Comment (colinbruce) [inline app/controllers/external_users/advocates/claims_controller.rb:22]: :+1:

**Comment by jsugarman** [comment]
see rebased, amended and extended version https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/pull/2625

**Pull request closed without merging**